### PR TITLE
[Tune] Remove legacy resources implementations in Runner and Executor.

### PR DESCRIPTION
**Pull request by xwjiang2010**
## Why are these changes needed?
This PR tries to unify our internal implementation to use PGF across Runner, Executor and Trial. Externally, if the resources specification is simple (i.e. it doesn't include `extra_*`, just straight CPU, GPU number), we still allow user to use Resources. This is, imo, a better user experience. If the resources specification is complicated with `extra_*`, an error will be raised to ask user to migrate.

This PR also migrates some legacy call sites to use PGF. During this migration, I noticed that the current implementation of PGF does not work when the trainer doesn't need any resources itself, but its workers need. Hence there is some update to PGF implementation.

**Note**
Below are bugs opened to clarify test behavior. All other tests are migrated and verified. Will make sure these issues are closed before merging in the PR.
https://github.com/ray-project/ray/issues/19952
https://github.com/ray-project/ray/issues/19985
https://github.com/ray-project/ray/issues/19959

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by krfricke — APPROVED** [review]
Generally looks great! Thanks for getting to this!

We just have to make tests pass and I have one comment about a removed property

**Review by Yard1 — COMMENTED** [review]
Great work!

File: python/ray/tune/schedulers/hb_bohb.py [hunk 1]
```diff
@@ -119,8 +119,8 @@ def choose_trial_to_run(self,
             scrubbed = [b for b in hyperband if b is not None]
             for bracket in scrubbed:
                 for trial in bracket.current_trials():
-                    if (trial.status == Trial.PENDING
-                            and trial_runner.has_resources_for_trial(trial)):
+                    if (trial.status == Trial.PENDING and trial_runner.
+                            trial_executor.has_resources_for_trial(trial)):
```
[the remark this answers is not in the document]
Comment (xwjiang2010) [reply python/ray/tune/schedulers/hb_bohb.py:123]: Tried it manually, this works. Maybe has something to do with the fact it's wrapped up in a ().

File: python/ray/tune/schedulers/hyperband.py [hunk 2]
```diff
@@ -304,8 +304,8 @@ def choose_trial_to_run(
             for bracket in sorted(
                     scrubbed, key=lambda b: b.completion_percentage()):
                 for trial in bracket.current_trials():
-                    if (trial.status == Trial.PENDING
-                            and trial_runner.has_resources_for_trial(trial)):
+                    if (trial.status == Trial.PENDING and trial_runner.
+                            trial_executor.has_resources_for_trial(trial)):
```
[the remark this answers is not in the document]
Comment (xwjiang2010) [reply python/ray/tune/schedulers/hyperband.py:308]: Tried it manually, this works. Maybe has something to do with the fact it's wrapped up in a ().

File: python/ray/tune/trial.py [hunk 3]
```diff
@@ -458,10 +438,6 @@ def remote_checkpoint_dir(self):
         logdir_name = os.path.basename(self.logdir)
         return os.path.join(self.remote_checkpoint_dir_prefix, logdir_name)
 
-    @property
-    def uses_placement_groups(self):
-        return bool(self.placement_group_factory)
-
```
[the remark this answers is not in the document]
Comment (xwjiang2010) [reply python/ray/tune/trial.py:464]: This seems an internal API to me. We should probably annotate them accordingly?
@PublicAPI, @DeveloperAPI etc..

File: python/ray/tune/tests/test_api.py [hunk 4]
```diff
@@ -277,12 +275,15 @@ def f(cpus, gpus):
...
+        # self.assertEqual(f(0, 0).status, Trial.TERMINATED)
 
         # Too large resource request
-        self.assertRaises(TuneError, lambda: f(100, 100))
-        self.assertRaises(TuneError, lambda: f(0, 100))
-        self.assertRaises(TuneError, lambda: f(100, 0))
+        # TODO(xwjiang): Re-enable after https://github.com/ray-project/ray/issues/19985.  # noqa
+        # self.assertRaises(TuneError, lambda: f(100, 100))
+        # self.assertRaises(TuneError, lambda: f(0, 100))
+        # self.assertRaises(TuneError, lambda: f(100, 0))
```
Comment (richardliaw) [inline python/ray/tune/tests/test_api.py:287]: Can we make sure these tests pass first?

File: python/ray/tune/tests/test_trial_runner_2.py [hunk 5]
```diff
@@ -133,43 +132,6 @@ def testFailureRecoveryEnabled(self):
         self.assertEqual(len(searchalg.errored_trials), 0)
         self.assertEqual(len(scheduler.errored_trials), 0)
 
-    def testFailureRecoveryNodeRemoval(self):
-        # Node removal simulation only works with resource requests
-        os.environ["TUNE_PLACEMENT_GROUP_AUTO_DISABLED"] = "1"
-
```
Comment (richardliaw) [inline python/ray/tune/tests/test_trial_runner_2.py:139]: is there a similar test that we can add?

**Review by richardliaw — COMMENTED** [review]
I think it makes more sense to remove the older implementation once we make sure our testing framework is in place (instead of commenting out tests to allow this change to go through)

File: python/ray/tune/tests/test_api.py (hunk 4 above)
Comment (xwjiang2010) [reply python/ray/tune/tests/test_api.py:287]: It's gonna take some time tho. Need that autoscaler API.

File: python/ray/tune/tests/test_trial_runner_2.py (hunk 5 above)
Comment (xwjiang2010) [reply python/ray/tune/tests/test_trial_runner_2.py:139]: `test_trial_requeue` is the one mimicking node failure.  It's currently failing. Will make sure that one passes before merging this in.

**Comment by xwjiang2010** [comment]
> I think it makes more sense to remove the older implementation once we make sure our testing framework is in place (instead of commenting out tests to allow this change to go through)

Yes, 
Below are bugs opened to clarify test behavior. All other tests are migrated and verified. Will make sure these issues are closed before merging in the PR.
#19952
#19985
#19959

File: python/ray/tune/ray_trial_executor.py [hunk 6]
```diff
@@ -189,7 +189,6 @@ def __init__(self,
         self._cached_actor_pg = deque(maxlen=1)
 
         self._avail_resources = Resources(cpu=0, gpu=0)
```
Comment (Yard1) [inline python/ray/tune/ray_trial_executor.py:191]: Should this be staying?

**Review by Yard1 — COMMENTED** [review]
Looks great! I am a little bit worried about commenting out tests, but I understand the reason. One thing I wonder is if we should wait on removing non-private methods and properties for now, and instead do a depreciation cycle.

File: python/ray/tune/ray_trial_executor.py (hunk 6 above)
Comment (xwjiang2010) [reply python/ray/tune/ray_trial_executor.py:191]: Seems like it's used in `debug_string`, which is printed out  in Tune loop. I find the information quite useful.
How about let's keep it here for now.

File: python/ray/tune/utils/placement_groups.py [hunk 7]
```diff
@@ -429,18 +463,18 @@ def get_full_actor_cls(self, trial: "Trial",
...
+            head_bundle = pg.bundle_specs[0].copy()
+            num_cpus = head_bundle.pop("CPU", 0)
+            num_gpus = head_bundle.pop("GPU", None)
+
+            # Only custom resources remain in `head_bundle`
+            resources = head_bundle
+        else:
+            num_cpus = None
+            num_gpus = None
+            resources = None
```
Comment (krfricke) [inline python/ray/tune/utils/placement_groups.py:477]: Does this work? Do we have a test for this path?

**Review by krfricke — COMMENTED** [review]
Intermediate review: Looks great so far! I'll check the tests later

File: python/ray/tune/utils/placement_groups.py (hunk 7 above)
Comment (xwjiang2010) [reply python/ray/tune/utils/placement_groups.py:477]: @rkooo567 This is the case we were discussing today. Basically actor itself doesn't need anything. But its child will require some non trivial resources. Please advice what is the right way to call PlacementGroup in this case.

File: python/ray/tune/utils/placement_groups.py [hunk 8]
```diff
@@ -104,10 +119,19 @@ class PlacementGroupFactory:
...
+                 bundles: List[Dict[str, Union[int, float]]],
                  strategy: str = "PACK",
                  *args,
                  **kwargs):
+        assert len(bundles) > 0, (
+            "Cannot initialize a PlacementGroupFactory with zero bundles.")
+        if not bundles[0]:
+            # This is when trainable itself doesn't need resources.
+            self._head_bundle_is_empty = True
+            bundles.pop(0)
```
Comment (rkooo567) [inline python/ray/tune/utils/placement_groups.py:131]: If you pop the first bundle, isn't it causing problem when you schedule actors on bundle index 0?

File: python/ray/tune/utils/placement_groups.py (hunk 7 above)
Comment (xwjiang2010) [reply python/ray/tune/utils/placement_groups.py:477]: https://sourcegraph.com/github.com/ray-project/ray/-/blob/python/ray/tune/integration/horovod.py?L256
The previous behavior is using legacy resources, which we are trying to remove here.

So it seems like pg and resources are not reaching feature parity?

Comment (xwjiang2010) [reply python/ray/tune/utils/placement_groups.py:477]: If it is not too much headache Core side, I would propose that we should reach feature parity by allowing actor using 0 resources.

File: python/ray/tune/utils/placement_groups.py (hunk 8 above)
Comment (xwjiang2010) [reply python/ray/tune/utils/placement_groups.py:131]: You could be right now that you explained to me what that bundle index 0 argument means.
Although I don't observe any issues when tests run through this branch; we should clean up the API semantics here. i.e. Applications should be able to specify the case where actor requires 0 resources.

File: python/ray/tune/utils/placement_groups.py (hunk 7 above)
Comment (rkooo567) [reply python/ray/tune/utils/placement_groups.py:477]: oh actually it reaches the feature parity. You can schedule actor/task with 0 resource requests. You just cannot create 0 resource bundle (which doesn't make sense).

File: python/ray/tune/trial.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (krfricke) [reply python/ray/tune/trial.py:476]: Let's remove it here

**Review by krfricke — APPROVED** [review]
Amazing, let's merge once tests passed

**Comment by krfricke** [comment]
Unfortunately one test is still failing: https://buildkite.com/ray-project/ray-builders-pr/builds/18819#7789d278-2104-415a-860d-d765eafea804/473-689

**Comment by xwjiang2010** [comment]
> Unfortunately one test is still failing: https://buildkite.com/ray-project/ray-builders-pr/builds/18819#7789d278-2104-415a-860d-d765eafea804/473-689

interesting. looking into it. It doesn't fail locally. let me rerun first.

**Comment by xwjiang2010** [comment]
All green. Not sure why that test becomes flaky. Tried multiple times locally, it's consistently passing .

**Pull request merged**